File: internal/extension/artifacts/worker/src/token.ts
/**
 * The verifying half of internal/extension/artifacts/token.go.
 *
 * Two implementations of one format is a liability, so the wire format is
 * kept deliberately dull -- base64url, SHA-256, newline-joined canonical
 * strings, no canonical JSON anywhere -- and testdata/vectors.json is
 * minted by the Go test and verified by the TypeScript one, so a change to
 * either side that breaks the other fails a test rather than a deploy.
 */

const ENCODER = new TextEncoder();

/** How far a signed request's timestamp may sit from this worker's clock. */
export const AUTH_SKEW_SECONDS = 300;

export interface Claims {
  id: string;
  exp: number;
}

/**
 * Returns a view onto a plain ArrayBuffer rather than ArrayBufferLike.
 * Uint8Array's usual inferred type widens to include SharedArrayBuffer,
 * which WebCrypto's BufferSource does not accept -- so the buffer is
 * allocated explicitly and filled, which is the same work and the type the
 * crypto calls actually want.
 */
function b64urlDecode(value: string): Uint8Array<ArrayBuffer> {
  const padded = value.replace(/-/g, "+").replace(/_/g, "/").padEnd(Math.ceil(value.length / 4) * 4, "=");
  const binary = atob(padded);
  const bytes = new Uint8Array(new ArrayBuffer(binary.length));
  for (let index = 0; index < binary.length; index += 1) bytes[index] = binary.charCodeAt(index);
  return bytes;
}

function b64urlEncode(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

async function hmacKey(secret: string): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", ENCODER.encode(secret.trim()), { name: "HMAC", hash: "SHA-256" }, false, [
    "sign",
    "verify",
  ]);
}

async function hex(bytes: ArrayBuffer): Promise<string> {
  return [...new Uint8Array(bytes)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

/**
 * Verify a read link's key for one artifact.
 *
 * Returns the claims or null, never a reason: separating "expired" from
 * "forged" tells whoever is probing which half to work on, and the holder
 * of a bad link can do nothing with either answer but ask for a new one.
 */
export async function verifyLink(secret: string, token: string, id: string, nowSeconds: number): Promise<Claims | null> {
  const separator = token.indexOf(".");
  if (separator < 0) return null;
  const payload = token.slice(0, separator);
  const signature = token.slice(separator + 1);

  let provided: Uint8Array<ArrayBuffer>;
  try {
    provided = b64urlDecode(signature);
  } catch {
    return null;
  }

  // crypto.subtle.verify is constant-time, so this leaks nothing about the
  // expected signature.
  const valid = await crypto.subtle.verify("HMAC", await hmacKey(secret), provided, ENCODER.encode(payload));
  if (!valid) return null;

  let claims: Claims;
  try {
    claims = JSON.parse(new TextDecoder().decode(b64urlDecode(payload))) as Claims;
  } catch {
    return null;
  }

  if (typeof claims.id !== "string" || !Number.isFinite(claims.exp)) return null;
  // Scope. Without it, a link to any artifact opens every artifact.
  if (claims.id !== id) return null;
  if (claims.exp <= nowSeconds) return null;
  return claims;
}

/**
 * Verify a signed publish or list request.
 *
 * The signature covers the verb, the path, the timestamp and a digest of
 * the body, so a captured header cannot be replayed onto another artifact,
 * another verb, or the same artifact with different content.
 */
export async function verifyRequest(
  secret: string,
  header: string | null,
  method: string,
  path: string,
  body: ArrayBuffer,
  nowSeconds: number
): Promise<boolean> {
  if (!header) return false;
  const parts = header.split(":");
  if (parts.length !== 3 || parts[0] !== "v1") return false;
  const [, stamp, signature] = parts;

  const at = Number(stamp);
  if (!Number.isFinite(at) || Math.abs(nowSeconds - at) > AUTH_SKEW_SECONDS) return false;

  const digest = await hex(await crypto.subtle.digest("SHA-256", body));
  const canonical = [method, path, stamp, digest].join("\n");

  let provided: Uint8Array<ArrayBuffer>;
  try {
    provided = b64urlDecode(signature);
  } catch {
    return false;
  }
  return crypto.subtle.verify("HMAC", await hmacKey(secret), provided, ENCODER.encode(canonical));
}

/**
 * Mint a read link's key, exactly as the Go client's MintLink does. The
 * index page uses it to hand out a working link per artifact. Same claim
 * order, same encoding, no whitespace: a Worker-minted key for the same
 * artifact and expiry is byte-identical to a Go-minted one, and the
 * vectors test holds both sides to that.
 */
export async function mintLink(secret: string, id: string, expSeconds: number): Promise<string> {
  const payload = b64urlEncode(ENCODER.encode(JSON.stringify({ id, exp: expSeconds })));
  const signature = await crypto.subtle.sign("HMAC", await hmacKey(secret), ENCODER.encode(payload));
  return `${payload}.${b64urlEncode(new Uint8Array(signature))}`;
}

export { b64urlEncode, b64urlDecode };

File: extension/artifacts/token.go
package artifacts

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// authHeader carries a signed publish or list request. Read links travel in
// the URL instead, because the URL is the thing being shared.
const authHeader = "X-Artifact-Auth"

// linkParam is the query parameter a read link's key travels in.
const linkParam = "k"

// authSkew is how far a signed request's timestamp may be from the worker's
// clock. It bounds replay without requiring the two to be in step; a
// laptop that has been asleep is the usual cause of a wider gap.
const authSkew = 5 * time.Minute

// claims is what a read link asserts. It is signed, not encrypted, so the
// holder can read it: that rules out putting the publishing account or
// session in here, since a link sent outside the team would carry them.
//
// The scope is one artifact. Without it a link to anything would open
// everything, which is the whole difference between a shared link and a
// shared password.
type claims struct {
	ID  string `json:"id"`
	Exp int64  `json:"exp"`
}

func b64(raw []byte) string {
	return base64.RawURLEncoding.EncodeToString(raw)
}

func decodeB64(encoded string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(encoded)
}

func mac(key, message []byte) []byte {
	h := hmac.New(sha256.New, key)
	h.Write(message)
	return h.Sum(nil)
}

// mintLink returns the key for one artifact, valid until exp.
//
// The signature covers the encoded payload rather than the claims struct,
// so the worker verifies the exact bytes it received and never has to
// re-encode JSON to check them. Field order, spacing and Go's map
// iteration therefore cannot break verification.
func mintLink(key []byte, id string, exp time.Time) (string, error) {
	if len(key) == 0 {
		return "", fmt.Errorf("no signing key")
	}
	if id == "" {
		return "", fmt.Errorf("no artifact id")
	}
	encoded, err := json.Marshal(claims{ID: id, Exp: exp.Unix()})
	if err != nil {
		return "", err
	}
	payload := b64(encoded)
	return payload + "." + b64(mac(key, []byte(payload))), nil
}

// verifyLink is the Go side of what the worker does per request. Nothing on
// the publish path needs it; it exists so the two implementations are
// checked against each other by test rather than by deployment.
//
// Every rejection is the same error. Telling a caller that a token was
// expired rather than forged tells an attacker which half to work on, and
// the holder of a bad link can do nothing with either answer but ask for
// another one.
func verifyLink(key []byte, token, id string, now time.Time) error {
	bad := fmt.Errorf("invalid or expired link")
	payload, signature, found := strings.Cut(token, ".")
	if !found {
		return bad
	}
	want, err := base64.RawURLEncoding.DecodeString(signature)
	if err != nil {
		return bad
	}
	if !hmac.Equal(want, mac(key, []byte(payload))) {
		return bad
	}
	raw, err := base64.RawURLEncoding.DecodeString(payload)
	if err != nil {
		return bad
	}
	var parsed claims
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return bad
	}
	if parsed.ID != id || parsed.Exp <= now.Unix() {
		return bad
	}
	return nil
}

// signRequest authenticates a publish or list call.
//
// The signature covers the method, the path, the timestamp and a digest of
// the body, so a captured header cannot be replayed against a different
// artifact, a different verb, or the same artifact with different content.
func signRequest(key []byte, method, path string, at time.Time, body []byte) string {
	digest := sha256.Sum256(body)
	stamp := strconv.FormatInt(at.Unix(), 10)
	canonical := strings.Join([]string{method, path, stamp, hex.EncodeToString(digest[:])}, "\n")
	return "v1:" + stamp + ":" + b64(mac(key, []byte(canonical)))
}
